[FLINK-7657] [Table] Adding logic to convert RexLiteral to expected SQL Date/Time/Timestamp classes

kmurra opened the pull request:
## What is the purpose of the change

This change fixes handling of date, time, and timestamp literal expressions.

A ClassCastException occurs if you have a date, time, or timestamp literal included in the where clause of a SQL query.  When Calcite trees are converted to an Expression, Literal values are not properly converted to the expected types by the RegNodeToExpressionConverter.

See https://issues.apache.org/jira/browse/FLINK-7657 for more information.

## Brief change log

- Added a method that takes a RexLiteral in the Literal object and applies the appropriate transforms to date & time related fields.
- Have a default fallback that works using the old behavior for other field types.  
    - Note that handling of other types are incorrect per the public documentation, but fixes are out of scope for this JIRA
- Added comments that explained handling of values (specifically why we're shifting them based on time zones)

## Verifying this change

This change added tests and can be verified as follows:
- Unit test cases were added covering changes
  - Added unit test case to RexProgramExtractorTest.scala to test that literal conversions are of the appropriate type.
  - Added unit test case to TableSourceTest.scala to confirm that received Expressions in a table source extending from FilterableTableSource do not cause exceptions and have expected values
- Note: Other unit test cases cover possible regressions in output values from the Literal->RexNode conversion

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/literals.scala:
```diff
@@ -103,13 +148,21 @@ case class Literal(value: Any, resultType: TypeInformation[_]) extends LeafExpre
-    val cal = Calendar.getInstance(Literal.GMT)
-    val t = date.getTime
-    // according to Calcite's SqlFunctions.internalToXXX methods
-    cal.setTimeInMillis(t + TimeZone.getDefault.getOffset(t))
```
Could you explain why `cal.setTimeInMillis(t + TimeZone.getDefault.getOffset(t))` did not produce the correct result? Calcite is also doing it like that.

kmurra replied:
The re-implemented adjustCalendar method is functionally the same as this when toTz is the UTC TimeZone.  Its just generalized to allow converting between arbitrary TimeZones so that I can re-use it in the RexLiteral to Expression conversion.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/literals.scala:
```diff
@@ -49,10 +50,51 @@ object Literal {
+      case _@SqlTimeTypeInfo.DATE =>
+        val rexValue = rexNode.getValueAs(classOf[DateString])
+        val adjustedCal = adjustCalendar(rexValue.toCalendar, TimeZone.getDefault)
+        new Date(adjustedCal.getTimeInMillis)
```
This can be simplified to `new Date(rexValue.toString)`.

kmurra replied:
Unfortunately that constructor is deprecated in Java 8, which is why I avoided using it.

kmurra replied:
Do you want me to use the deprecated constructor or leave this as-is?

kmurra commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/CheckExpressionsTableSource.scala:
```diff
@@ -0,0 +1,65 @@
+  * @param assertions A set of assertions as a function reference
+  * @param pushedDown Whether this has been pushed down/
+  */
+class CheckExpressionsTableSource(typeInfo: RowTypeInfo,
```
I'll look at doing that.  It was my initial approach, but when I saw the potential set of test cases that would impact, I decided to do something more conservative.  Still do-able.

kmurra commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/TableTestBase.scala:
```diff
@@ -200,10 +201,30 @@ case class BatchTableTestUtil() extends TableTestUtil {
     printTable(tableEnv.sqlQuery(query))
   }
 
+  def verifyExpressionProjection(fields: Seq[(String, TypeInformation[_])],
```
I will move this.

kmurra commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/literals.scala:
```diff
@@ -49,10 +50,51 @@ object Literal {
     case sqlTimestamp: Timestamp => Literal(sqlTimestamp, SqlTimeTypeInfo.TIMESTAMP)
   }
+
+  private[flink] def apply(rexNode: RexLiteral): Literal = {
```
I'll commit to using your standards for the code-base.  However, allow me to voice a disagreement here:

The Literal class does the conversion from the Literal back to the RexLiteral.  Having this logic specifically in RexNodeToExpressionConverted means the RexLiteral-to-Literal logic is physically split from the Literal-to-RexLiteral logic.  This makes it slightly easier for a contributor to make a change in one side of the conversion without accounting for the other.  In particular, the date adjustments here become harder to understand since the context is split between two different files.

kmurra commented:
Regarding the time zones, I think I understand your argument here.  Is there anything in particular that you would want me to change overall that you haven't already outlined to account for that? I do want to document why we're doing any conversions of time zones since it took me some amount of time to understand why it was being done (it looked incorrect to myself and several other developers on first glance).

Also, I noticed that the Calcite fromCalendarFields simply take the fields directly from the Calendar, so making time-zone adjustments are unnecessary after I made the changes to toRexNode.  I'll fix that as well for my next commit.

kmurra commented:
The biggest change here is in the test cases -- I generalized the test table source to have some basic filtering logic and allow for generic datasets.

I moved the Literal build logic to the RexNodeToExpressionConverter.visitLiteral.  I also rewrote several of the conversion methods to more closely align with the intended behavior of the code - that we're preserving the string values of the various time-related literals in the local timezone.  This made a bunch of the epoch-millisecond modifications go away.

twalthr commented:
Thanks for the update @kmurra. I will go over the changes a last time and merge this.
